src/Button/Button.tsx

```diff
@@ -136,11 +137,21 @@ function getHoverBackgroundColor({
 // I was able to get guarantees to work, but only with very cryptic errors. I
 // decided it'd be best, for the time being, to `throw` if we use things
 // incorrectly.
-interface Props
-  extends React.DetailedHTMLProps<
-    React.ButtonHTMLAttributes<HTMLButtonElement>,
-    HTMLButtonElement
-  > {
+interface Props {
+  /**
+   * Lets you customize how you're going to render this component. You can
+   * either one of two things:
+   *
+   * 1. A string representing the type of element you want rendered, or
+   * 2. A render function. The render function will be called with the props
+   *    that need to be passed to the element. You can use this to change the
+   *    element being rendered or to add your own props that aren't included in
+   *    this component.
+   *
+   * @default "button"
+   */
+  as?: string | ((props: any) => ReturnType<React.FC>);
```

Comment:
@justinanastos 	let's thread this (responding to https://github.com/apollographql/space-kit/pull/70#issuecomment-521792084)

> What about when we want to use `partAs` 

What problems would be created by using `as` like this `<Button as={NavLink} to="/" />` that doesn't exist when used like this `<Button as={(props) => <Link {...props} to="some-route" />}>` wrt `partAs`?

Reply:
ah, is it typing? having trouble getting ts to not yell at me trying to add `partAs`

maybe more trouble than it's worth :p

Reply:
oh oh I got it working  

does this help?

![image](https://user-images.githubusercontent.com/743976/63127270-e7125180-bf7f-11e9-9744-8a502a694abf.png)

Reply:
First, like I said, I don't want to add new props to the component because we use the `as` prop. I'm trying to keep Space Kit's APIs simple; this adds this will now sometimes have a side-effect of adding ALL the props of your `as` component to the Button. That doesn't feel like a good API to me.

This also will require that we always use a non-anonymous function or rendering will unmount/mount on every render because you're using `React.createElement` (via JSX), where I was intentionally using a render function (meaning _not_ a `React.createElement`).

Going back to `partAs`, now we're going to say that "if you use `as` you use one API, and when you use `partAs` you use a different API." I'd rather keep this logic the same for `as` and `partAs`.

Reply:
When I say a "different API", I mean props from the `<Button>` will be spread into `as`. When using `partAs` you have to call the anonymous function (which will cause a mount/unmount on every render if we're using `React.createElement`) to add props.

Reply:
@trevorblades Can you give more than "I prefer it"? Like I said here, we are intentionally limiting the number of props on the component. This will add hundreds _sometimes_ which feels like a bad API.

Reply:
I'm not too bothered by the exposing of `as` component props on the `Button` component. The JSX reads better IMO, and as far as I can tell, doesn't impact the user negatively. In the following example, `A` looks nicer to me:

```js
// A
<Button as={Link} to="/path1" />
<Button as={Link} to="/path2" />
<Button as={Link} to="/path3" />

// B
<Button as={props => <Link to="to="/path1" {...props} />} />
<Button as={props => <Link to="to="/path2" {...props} />} />
<Button as={props => <Link to="to="/path3" {...props} />} />
```

Assuming we went with the render props method, would there be any props that are passed to the button (ie. `feel`) that we wouldn't want to pass to the underlying component (`Link` in this case)? Would I have to write my render function like this?

```js
<Button
  feel="raised"
  as={({feel, ...props}) => <Link to="/path1" {...props} />}
>
  Click me
</Button>
```

Using `React.createElement` in the `Button` component would give us control over what props are eventually passed to the underlying component.

Reply:
If I looked at the `Button` code in example `A` then I'd assume that `to` was a prop that exists on `Button`. I'd go to the docs and I wouldn't see it and I wouldn't know why. It's that misdirection that I want to avoid.

Second, we can't use `React.createElement` because we use emotion's `css`. I tried automatically passing the props through, including the `css`, and I couldn't make it work as we'd expect.

Lastly, if we use `React.createElement` (or `jsx` provided by emotion) to render a component you pass to `as`, that means we can never use a render function because this will mount/unmount the element on every render. This is why react router provides a `render` and `component` prop on `Route`.

I think I'm going to stick with the pattern from the startdust article. Also, something I missed is that they support passing `as={<a href="" />}`. I don't know if I can support that and use `css` and have `css` allowed to be overridden. I'll check that out tomorrow.

Thanks for all the feedback and thank you @cheapsteak for getting those types to work. I'm super impressed!

Reply:
If it matters, the pattern I proposed comes _from_ stardust   

https://github.com/stardust-ui/react/blob/master/docs/src/pages/Composition.mdx#unhandled-props--dom-attributes

![image](https://user-images.githubusercontent.com/743976/63131588-45ddc800-bf8c-11e9-9639-97a40fd0432a.png)

Reply:
I've said it in 3 message and I'd really like you directly respond to this @cheapsteak : if `React.createElement` is called an inline function it'll mount/unmount on every render because it sees you using a different component. To my knowledge, and I tried to go this direction from the start, there's no way to know if something is an SFC or a render function. Again, why `Route` in react-router has a `Component` prop and a `render` prop that do different things (this is over a year old but explains it well, also it was the first result when I googled it https://medium.com/@ilovezcd/difference-between-render-and-component-prop-on-react-router-v4-368ca7fedbec)

Reply:
In short, if we're going to agree we can use these forms; I don't know how to implement it:

```jsx
<Button as="button" />
```
```jsx
<Button as={<NavLink to="/"} />
```

```jsx
<Button as={NavLink} to="/" />
```

```jsx
<Button as={(props) => <NavLink {...props} to="/" />} />
```

Options 1-3 can use `React.createElement` (really emotion's `jsx` because this has to support emotion) and try to spread props. The 4th option has to be different because `React.createElement` will unmount on every render. That's not an acceptable trade-off to save 20 keystrokes when you create `<LinkButton>` component in each project that consumes this.

It also, like I keep saying, does not force us into the pit of success. We're providing an API that can very easily cause undesired and confusing behavior.

Reply:
What is the practical consequence of `<Button as={(props) => <NavLink {...props} to="/" />} />` unmounting on re-render? Also why would someone want to write `<Button as={(props) => <NavLink {...props} to="/" />} />` if they could have written `<Button as={NavLink} />`

This feels like a situation of letting the perfect be an enemy of the good

Reply:
- I believe there isn't a significant use case for `<Button as={(props) => <NavLink {...props} to="/" />} />` if we supported `<Button as={NavLink} to="/"/>`
- I believe for those situations where we need `<Button as={(props) => <NavLink {...props} to="/" />} />`, the performance cost will be negligible  
- I believe the potential (and imo minor) confusion caused by seeing a "to" prop exist on a "Button" component when it has `as={NavLink}` will not be a hindrance for the calibre of developers who are likely to work with this component library (Apollo devs)
- I believe there's value (albeit also minor) in following patterns established by styled-components, semantic-ui-react, stardust
- I believe developers who would work with this lib would prefer the ergonomics of `<Button as={NavLink} to="/"/>` (admittedly data point of 3 so far and only 2 in favour of it)


On those bases of belief, I believe `<Button as={NavLink} to="/"/>` would be a relatively more pragmatic API, but I have no way of supporting that besides intuition, and accept if your intuition differs from mine. At the end of the day, they would both solve the problem, and I think someone (I think you) will need to put their foot down :)

Reply:
I'm fine going in that direction @cheapsteak ; I'm trying to map this to how we'll use it for `labelAs` for `<Input>`, which we also need right now.

I guess we can say "don't use a render prop function" and leave it at that.

Reply:
@cheapsteak Can you take a shot at updating the props above to work with a default element of "button" as well?

Reply:
Got the typings to work with quite a bit of help from a typescript guru ( shouting out to @jephuff)

https://github.com/apollographql/space-kit/blob/53c5a73d6434ce0b60cd45acbe7fdf9486490e2e/src/Button/Button.tsx

still a bit sloppy and didn't spend much time with variable naming, but maybe you could work with this?